### PR TITLE
Smart Charging: Expand add profile capabilities

### PR DESCRIPTION
**Pull request by couryrr-afs**
## Describe your changes

- Added logic to replace existing non-external constraint charging profiles
- Added logic to verify TX Default charging profile evse placement.
- Added test to cover required callbacks.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] If OCPP 2.0.1: I have updated the [OCPP 2.0.1 status document](https://github.com/EVerest/libocpp/tree/main/doc/ocpp_201_status.md)
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by couryrr-afs** [comment]
@shankari this is ready for your review.

File: lib/ocpp/v201/smart_charging.cpp
Comment (louisg1337) [inline lib/ocpp/v201/smart_charging.cpp:333]: Just for documentation purposes, I think leaving a comment that this relates to K01.FR.05 could be helpful.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (louisg1337) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1029]: Not sure if it is necessary, but to be thorough to the K01FR05 tests, it could be worth adding in the opposites of these tests:

1. Same id does **NOT** exist and `ChargingStationExternalContraints`
2. Same id does **NOT** exist and **NO** `ChargingStationExternalContraints`

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (louisg1337) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:983]: Another test here could potentially be helpful:

1. `TxDefaultProfile` and same stack level **DOES** exist

Comment (couryrr-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:983]: This test is unique in the sense that we are only testing that add is placing the profile in the correct location. The flow would normally have the charging profiles going through a validation step before hand which would catch this sort of incorrect state. If you take a look in the tests for `DuplicateTxDefaultProfileFound` that is where this is validated.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (couryrr-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:1029]: For case 1 I believe that is covered with the normal `add_profile` test. Keeping in mind that this is only the ability to add profiles. It does not cover the validation of profiles.

This test is specific to the case of `K01FR05` with that being the lens to look at this through case 2 would not help in catching regressions to `ChargingStationExternalContraints`.

File: tests/lib/ocpp/v201/test_charge_point.cpp
Comment (shankari) [inline tests/lib/ocpp/v201/test_charge_point.cpp:145]: is the callback validity actually a test case for `K01FR02`? Why? It doesn't seem to be related to the requirements in the spec.

**Review by shankari — APPROVED** [review]
This is a fairly simple change and my comments are primarily focused on improved documentation, so I am approving this. Please do address my comments before sending it off to the community for merging. Documentation clarity is just as important, particularly in open source projects, where contributors can join and leave, as code.

File: tests/lib/ocpp/v201/test_charge_point.cpp
Comment (drmrd) [reply tests/lib/ocpp/v201/test_charge_point.cpp:145]: @couryrr-afs: I agree with @shankari that this could use some context. Could you move the `all_callbacks_valid` tests that involve `set_charging_profiles_callback` to the top of these tests (if they're not there already) and include the following comment:

```c++
/*
 * K01.FR.02 states
 *
 *     "The CSMS MAY send a new charging profile for the EVSE that SHALL be used
 *      as a limit schedule for the EV."
 *
 * When using libocpp, a charging station is notified of a new charging profile
 * by means of the set_charging_profiles_callback. In order to ensure that a new
 * profile can be immediately "used as a limit schedule for the EV", a
 * valid set_charging_profiles_callback must be provided.
 *
 * As part of testing that K01.FR.02 is met, we provide the following tests that
 * confirm an OCPP 2.0.1 ChargePoint with smart charging enabled will only
 * consider its collection of callbacks valid if set_charging_profiles_callback
 * is provided.
 */
```

Following those tests, we can add a comment like the following:
```c++
/*
 * For completeness, we also test that all other callbacks are checked by
 * validate_all_profiles.
 */
```

The remaining `validate_all_profiles` tests could immediately follow this comment but with references to K01.FR.02 removed from their names.

Comment (drmrd) [reply tests/lib/ocpp/v201/test_charge_point.cpp:145]: I could see moving the other tests to a different location in the future but think it's fine to have them here for now. If other validity tests are relevant for functional requirements in the other functional blocks (as is highly likely), they can be moved around as needed. For now, though, it's nice to keep them grouped together.

File: lib/ocpp/v201/charge_point.cpp
Comment (drmrd) [reply lib/ocpp/v201/charge_point.cpp:37]: @shankari: The `security_event_callback` is defined as non-optional in `charge_point.hpp`:

https://github.com/EVerest/libocpp/blob/77f5b2a6181d94c2e94af790e4c5e61435f0b919/include/ocpp/v201/charge_point.hpp#L146-L152

It looks to me like @gberardi-pillar's change is meant to correct for this.

Comment (drmrd) [reply lib/ocpp/v201/charge_point.cpp:37]: That said, the only usage of `security_event_callback` in `libocpp` appears to assume that it's optional:

https://github.com/EVerest/libocpp/blame/587d491589f944c8a99ce8fe56ece612e061f385/lib/ocpp/v201/charge_point.cpp#L1878-L1880

So perhaps it would be better to change the type of `security_event_callback` to `std::optional<std::function>` instead of add it to the validate method?

File: lib/ocpp/v201/charge_point.cpp
Comment (drmrd) [reply lib/ocpp/v201/charge_point.cpp:59]: These appear to be optional callbacks. @couryrr-afs: I think this hunk should be removed from the PR along with the corresponding unit tests for `all_callbacks_valid`.

File: lib/ocpp/v201/charge_point.cpp
Comment (couryrr-afs) [reply lib/ocpp/v201/charge_point.cpp:37]: I have added these comments.

File: doc/ocpp_201_status.md
Comment (shankari) [inline doc/ocpp_201_status.md:1247]: These are not addressed in this PR, right?

**Review by shankari — APPROVED** [review]
One more clarifying question....

File: doc/ocpp_201_status.md
Comment (couryrr-afs) [reply doc/ocpp_201_status.md:1247]: These were addressed with the check [here](https://github.com/EVerest/libocpp/blob/b67951cebbff4de26456e4575e3ff6e84c4630ed/lib/ocpp/v201/smart_charging.cpp#L319)

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:323]: Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.

K01.FR.05 says that a profile with the same id should replace the existing profile. There is no mention of if the new profile is for the same EVSE etc.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:339]: I see that this is used for testing so can I assume this is temporary? I don't see a other reason to expose all the profiles to the outside since this class is responsible for the implementation.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (marcemmers) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:935]: Building on that last comment about `get_profiles()`, I don't see the point in testing it like this. From the test names they seem like good tests but I expect this to be tested using some kind of composite results in the future since that should be the primary interface of this.

**Review by marcemmers — COMMENTED** [review]
I have some comments/questions. Not sure if there are any actual changes needed but would like some answers at least before approving.

File: lib/ocpp/v201/smart_charging.cpp
Comment (drmrd) [reply lib/ocpp/v201/smart_charging.cpp:323]: > Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.

I believe the separate `station_wide_charging_profiles` variable was created for clarity, since various expected behaviors involving profiles with EVSE ID 0 are different than for other EVSE IDs. (@couryrr-afs please correct me if I'm wrong here.)

Comment (drmrd) [reply lib/ocpp/v201/smart_charging.cpp:323]: > K01.FR.05 says that a profile with the same id should replace the existing profile. There is no mention of if the new profile is for the same EVSE etc.

+1. If they're not already present, there should be test cases here for ensuring that adding a replacement charging profile (that is, one with the same ID as an existing profile) for a different EVSE than the existing profile results in the new profile being added and the old one being deleted.

Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:323]: > > Why do we have a separate store for station wide profiles and evse related charging profiles? For this use case it should be a single list I think. This would also be more in line with how the database is going to look like.
> 
> I believe the separate `station_wide_charging_profiles` variable was created for clarity, since various expected behaviors involving profiles with EVSE ID 0 are different than for other EVSE IDs. (@couryrr-afs please correct me if I'm wrong here.)

I seem to recall a PR comment asking for them to be separated, although I am having trouble finding it now.

But yes, it was because the concept of a profile on EVSE with ID 0 is handled differently and so the code clarity of separating it explicitly made sense.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (gberardi-pillar) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:935]: I'm unclear if you are saying that you want the composite results/future for this PR? For now, we have a straightforward and simple way to verify that profiles are being added, and if this behavior can be tested a different way in the future, we can iterate/incrementally build to that state.

File: lib/ocpp/v201/smart_charging.cpp
Comment (gberardi-pillar) [reply lib/ocpp/v201/smart_charging.cpp:339]: I agree that testing behavior is ideal rather than testing implementation. At the moment, we have a way to confirm that adding profiles works as expected. If, in the future, there is a better way to test the behavior without having this interface, we could always revisit, of course.

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp:323]: K01.FR.05 says:

> When a SetChargingProfileRequest with an already known ChargingProfile.id is received AND the existing ChargingProfile does NOT have chargingProfilePurpose = `ChargingStationExternalConstraints`, The Charging Station SHALL replace the existing ChargingProfile with the one specified.

When [working on K01.FR.05](https://github.com/US-JOET/libocpp/pull/11) it wasn't clear whether or not the charging profile should be replaced when they're in different places. Station-wide profiles are treated differently from EVSE-specific profiles - they're meant to be apply to all EVSEs. K01.FR.14 also says "The Charging Station SHALL apply, but not copy, this profile to all EVSEs" for profiles on EVSE 0. So that implies that station-wide profiles should be independent from the EVSE-specific profiles. This made it seem odd to replace a station-wide profile that applies to all EVSEs with a profile that doesn't or vice-versa. Should those profiles be replaced then? If so, I think it makes sense to have a test case.

Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp:323]: As discussed in the working group call, requirement K01.FR.05 is pretty clear that the charging profile id should be unique for the whole charge point. Therefore it indeed should be replaced and I agree adding a test is the best course of action.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp:339]: Yep understood, no need to block this for now but I think it's good to look at this again in the future.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (marcemmers) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp:935]: I don't expect any changes based on this comment. This was just elaborating on my other comment about exposing the `get_profiles()` function. I agree that we can iterate this in the future once we have some more composite schedule functions.

File: tests/lib/ocpp/v201/test_smart_charging_handler.cpp
Comment (couryrr-afs) [inline tests/lib/ocpp/v201/test_smart_charging_handler.cpp:992]: Based on the need for profile_ids to be unique. This test seems to be validating an invalid case.

Comment (christopher-davis-afs) [reply tests/lib/ocpp/v201/test_smart_charging_handler.cpp]: With the latest update we've removed this test.

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp]: We refactored how we store profiles and simplified the logic, and added new tests to ensure profile IDs are unique in all cases. We also added some new logic and tests to `validate_profile()` in order to ensure we don't replace `ChargingStationExternalConstraints` profiles.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:354]: I think we should be able to do this without needing an exception. Maybe check using `.count()` or `.find()` to check if the `STATION_WIDE_ID` is present.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [reply lib/ocpp/v201/smart_charging.cpp]: Looks like a good change, makes sense to use the map to also store the station wide profiles.

File: lib/ocpp/v201/smart_charging.cpp
Comment (marcemmers) [inline lib/ocpp/v201/smart_charging.cpp:21]: Do you really need all of these includes here? `vector` is already used in the header file so it must be included somehow. `<memory>` is directly included in the header file so not needed here.

**Review by marcemmers — APPROVED** [review]
Except for 2 small comments this looks good. The comments are relatively minor and I can accept to merge this as is and fix those later if you want to.

File: lib/ocpp/v201/smart_charging.cpp
Comment (christopher-davis-afs) [reply lib/ocpp/v201/smart_charging.cpp]: Sorry, clangd adds imports very aggressively. Removed 👍🏾

**Pull request merged**
